cpp/src/ffi_go/ir/serializer.cpp

```diff
@@ -0,0 +1,153 @@
+#include "serializer.h"
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include <clp/components/core/src/ffi/encoding_methods.hpp>
+#include <clp/components/core/src/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/components/core/src/ffi/ir_stream/encoding_methods.hpp>
+
+#include <ffi_go/defs.h>
+#include <ffi_go/ir/LogTypes.hpp>
+#include <ffi_go/LogTypes.hpp>
+
+namespace ffi_go::ir {
+using namespace ffi;
+using namespace ffi::ir_stream;
+
+namespace {
+    /**
+     * Generic helper for ir_serializer_serialize_*_log_event
+     */
+    template <class encoded_variable_t>
+    auto serialize_log_event(
+            StringView log_message,
+            epoch_time_ms_t timestamp_or_delta,
+            void* ir_serializer,
+            ByteSpan* ir_view
+    ) -> int {
```

Comment:
Seems like we don't check whether the input ptrs are null. Shall we add the check?
If yes, we should also add the check wherever a passed-in ptr gets dereferenced.